### DISCUSSION TITLE
langiumLangiumDocuments.getOrCreateDocument return all properties except one properties

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1271] [Q&A] [answered]
my requirement  ast root node is Ld like below second screenshot,when i call langiumLangiumDocuments.getOrCreateDocument(URI.parse(rootUri)) method,it return all properties(coils,ldnodes,locks) like below first screenshot except 'diagram',i want to know why(note:currently ,the corresponding file is an empty file)
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/a86f0f00-cebe-4a07-b7e8-79207b69f90d)
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/3c13f511-6991-4deb-bcb6-9176c5070897)
![image](https://github.com/eclipse-langium/langium/assets/26995040/88131c5b-486c-4f8b-a07e-38b271e60327)

**Comment by msujew** [comment]
@zrwsmd Does your grammar always expect an element of `VirtualDiagram` to be parsed? Since you're parsing an empty file, so Langium isn't able to actually create `VirtualDiagram` node from the input. Arrays are always available, as Langium can easily generate a default value for them.

Langium gives no guarantee that the AST is valid in the presence of parser errors.

**Comment by zrwsmd** [comment]
there is no parse error
i guess it because document is empty, the array properties in value are initialized with empty arrays [], but the object properties, such as diagram remain undefined,but how could it not be undefined?my requirement is when i drag a node and save it ,corresponding file  have text content,but when i create a node like below screenshot,at moment this.state.diagramRoot is still undefined,how could i assign it
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/258245aa-d9bb-49c1-8c39-3d8b052564c9)

**Reply by msujew** [reply, marked as answer]
In the second screenshot you've posted, the parse result clearly contains a parser error.

As for the requirement you're talking about: You could always create a synthetic diagramRoot element and put it into the state object. But I don't really see why this is necessary in the first place. If this is effectively modelling a graph  wouldn't the added new node immediately become the root node in the case of an empty file? Why the need for a separate root node?

**Reply by zrwsmd** [reply]
very thanks ,my eyes are blurred and I did not see it,i always thought there is no error
